pkg/processor/metricfrequencyprocessor/sieve.go

```diff
@@ -0,0 +1,209 @@
+package metricfrequencyprocessor
+
+import (
+	"math"
+	"sort"
+	"time"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+const (
+	float64EqualityThreshold = 1e-9
+	minPointAccumulationTime = time.Minute * 15
+	category1ReportFrequency = time.Minute * 5
+	category2ReportFrequency = time.Minute * 2
+	category3ReportFrequency = time.Second * 30
+	safetyInterval           = time.Second * 1
+
+	iqrAnomalyCoef            = 1.5
+	variationIqrThresholdCoef = 4
+)
+
+// metricSieve removes data points from MetricSlices that would be reported more often than preset
+// frequency for a given category.
+// For metric sieve, there are three categories of metrics:
+// 1) Constant metrics
+// 2) Low info metrics - i.e. no anomaly in terms of iqr and low variation
+// 3) All other metrics
+type metricSieve struct {
+	metricCache  *metricCache
+	lastReported map[string]pdata.Timestamp
+}
+
+func newMetricSieve() *metricSieve {
+	return &metricSieve{
+		metricCache: newMetricCache(),
+		lastReported: make(map[string]pdata.Timestamp),
+	}
+}
+
+// Sift removes data points from MetricSlices of the metric argument according to specified strategy.
+// It returns true if the metric should be removed.
+func (fs *metricSieve) Sift(metric pdata.Metric) bool {
+	switch metric.DataType() {
+	case pdata.MetricDataTypeDoubleGauge:
+		return fs.siftDropGauge(metric)
+	default:
+		return false
+	}
+}
+
+func (fs *metricSieve) siftDropGauge(metric pdata.Metric) bool {
+	metric.DoubleGauge().DataPoints().RemoveIf(fs.siftDataPoint(metric.Name()))
+
+	return metric.DoubleGauge().DataPoints().Len() == 0
+}
+
+func (fs *metricSieve) siftDataPoint(name string) func(pdata.DoubleDataPoint) bool {
+	return func(dataPoint pdata.DoubleDataPoint) bool {
+		cachedPoints := fs.metricCache.List(name)
+		fs.metricCache.Register(name, dataPoint)
+		lastReported, exists := fs.lastReported[name]
+		if !exists {
+			fs.lastReported[name] = dataPoint.Timestamp()
+			return false
+		}
+		earliest := earliestTimestamp(cachedPoints)
+		cachedPoints[dataPoint.Timestamp()] = dataPoint.Value()
+
+		if metricRequiresSamples(dataPoint, earliest) {
+			fs.lastReported[name] = dataPoint.Timestamp()
+			return false
+		}
+
+		if pastCategoryFrequency(dataPoint, lastReported, category1ReportFrequency) {
+			fs.lastReported[name] = dataPoint.Timestamp()
+			return false
+		}
+
+		if isConstant(dataPoint, cachedPoints) {
+			return true
+		}
+
+		if pastCategoryFrequency(dataPoint, lastReported, category2ReportFrequency) {
+			fs.lastReported[name] = dataPoint.Timestamp()
+			return false
+		}
+
+		if isLowInformation(cachedPoints) {
+			return true
+		}
+
+		if pastCategoryFrequency(dataPoint, lastReported, category3ReportFrequency) {
+			fs.lastReported[name] = dataPoint.Timestamp()
+			return false
+		}
```

Comment:
This section I believe needs a comment. What are 3 categories that are used here?